nncf/experimental/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,694 @@
+
+    def get_all_nodes(self) -> List[SearchGraphNode]:
+        """
+        Returns list of all graph nodes.
+        """
+        all_nodes = []
+        for node_key, node_attrs in self._nx_graph.nodes.items():
+            all_nodes.append(SearchGraphNode(node_key, node_attrs))
+        return all_nodes
+
+    def get_next_nodes(self, node_key: str) -> List[SearchGraphNode]:
+        """
+        Returns consumer nodes of provided node key.
+
+        :param node_key: Producer node key.
+        :return: List of consumer nodes of provided node.
+        """
+        next_node_keys = self._nx_graph.succ[node_key]
+        return [self.get_node_by_key(node_key) for node_key in next_node_keys]
+
+    def set_node_attr(self, node_key: str, name_attr: str, value_attr: str):
+        """
+        Set value of attribute by name for a given node with the same key.
+        """
+        self._nx_graph.nodes[node_key][name_attr] = value_attr
+
+    def get_prev_nodes(self, node_key: str) -> List[SearchGraphNode]:
+        """
+        Returns producer nodes of provided node key.
+
+        :param node_key: Consumer node key.
+        :return: List of producers nodes of provided node.
+        """
+        prev_node_keys = self._nx_graph.pred[node_key]
+        return [self.get_node_by_key(node_key) for node_key in prev_node_keys]
+
+    def get_prev_edges(self, node_key: str) -> Dict[str, Any]:
+        return self._nx_graph.pred[node_key]
+
+    def get_next_edges(self, node_key: str) -> Dict[str, Any]:
+        return self._nx_graph.succ[node_key]
+
+    def _insert_dummy_node(self, node_key: str):
+        # pylint: disable=protected-access
+        next_nodes = deepcopy(self._nx_graph._succ[node_key])
+        dummy_node_attr = {SearchGraph.IS_DUMMY_NODE_ATTR: True}
+        node_key_attr = deepcopy(self._nx_graph._node[node_key])
+        dummy_node_key = node_key + SearchGraph.DUMMY_POSTFIX
+        node_key_attr.update(dummy_node_attr)
+        self._nx_graph.add_node(dummy_node_key, **node_key_attr)
+
+        edge_attrs = None
+        for next_node_key in next_nodes:
+            edge_attrs = self._nx_graph.get_edge_data(node_key, next_node_key)
+            self._nx_graph.remove_edge(node_key, next_node_key)
+            self._nx_graph.add_edge(dummy_node_key, next_node_key, **edge_attrs)
+        self._nx_graph.add_edge(node_key, dummy_node_key, **edge_attrs)
+
+    def get_nx_graph(self) -> nx.DiGraph:
+        """
+        Returns internal representation of SearchGraph as a networkx graph.
+        """
+        return self._nx_graph
+
+def get_search_graph(original_graph: PTNNCFGraph) -> SearchGraph:
+    """
+    Returns a transformed representation of the network graph for blocks searching.
+    """
+    nx_merged_graph = get_merged_original_graph_with_pattern(original_graph.get_nx_graph_copy())
+    sgraph = SearchGraph(nx_merged_graph)
+    return sgraph
+
+def get_merged_original_graph_with_pattern(orig_graph: nx.DiGraph):
```

Comment:
Please add type hints here and below for returns parameters

Reply:
Done.